File: mobile-app/lib/core/vpn/vpn_log_redaction.dart
import 'dart:convert';

/// Маскирование типичных секретов Xray/VLESS в строках для debug/logcat.
abstract final class VpnLogRedaction {
  static final RegExp _uuid = RegExp(
    r'[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}',
    caseSensitive: false,
  );

  static final RegExp _quotedSensitive = RegExp(
    r'"(publicKey|privateKey|password|pbk|shortId)"\s*:\s*"[^"]*"',
    caseSensitive: false,
  );

  /// Идентификатор пользователя VLESS и др. в outbound.
  static final RegExp _idField = RegExp(
    r'"id"\s*:\s*"[^"]+"',
    caseSensitive: false,
  );

  static String redactSensitiveJson(String text) {
    var s = text.replaceAllMapped(
      _quotedSensitive,
      (m) => '"${m.group(1)}":"<redacted>"',
    );
    s = s.replaceAll(_idField, '"id":"<redacted>"');
    return s.replaceAll(_uuid, '<redacted>');
  }

  static String redactForLog(Object? value) {
    if (value == null) return 'null';
    final str = value is String ? value : jsonEncode(value);
    return redactSensitiveJson(str);
  }

  static String previewRedacted(String text, int maxLen) {
    final r = redactSensitiveJson(text);
    if (r.length <= maxLen) return r;
    return '${r.substring(0, maxLen)}…';
  }
}

File: mobile-app/android/app/src/main/kotlin/com/granivpn/mobile/VpnLogRedaction.kt
package com.granivpn.mobile

/**
 * Убирает из строк, попадающих в logcat, типичные секреты Xray/VLESS (UUID, ключи).
 * Не заменяет полноценный парсер JSON — достаточно для диагностических превью.
 */
object VpnLogRedaction {
    private val uuidRegex = Regex(
        "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}",
        RegexOption.IGNORE_CASE,
    )

    private val quotedSensitiveKey = Regex(
        "\"(publicKey|privateKey|password|pbk|shortId)\"\\s*:\\s*\"[^\"]*\"",
        RegexOption.IGNORE_CASE,
    )

    fun redactSensitiveJson(text: String): String {
        var s = quotedSensitiveKey.replace(text) { m ->
            "\"${m.groupValues[1]}\":\"<redacted>\""
        }
        s = uuidRegex.replace(s) { "<redacted>" }
        return s
    }

    fun previewRedacted(text: String, maxLen: Int): String {
        val r = redactSensitiveJson(text)
        return if (r.length <= maxLen) r else r.take(maxLen) + "…"
    }

    fun describeMethodCall(method: String, arguments: Any?): String {
        return when (method) {
            "connect" -> describeConnectArgs(arguments)
            else -> {
                val suffix = when (arguments) {
                    is Map<*, *> -> "keys=[${arguments.keys.joinToString(",")}]"
                    null -> "args=null"
                    else -> "argsType=${arguments.javaClass.simpleName}"
                }
                "method=$method $suffix"
            }
        }
    }

    private fun describeConnectArgs(arguments: Any?): String {
        if (arguments !is Map<*, *>) {
            return "method=connect argsType=${arguments?.javaClass?.simpleName ?: "null"}"
        }
        val keys = arguments.keys.joinToString(",")
        val config = arguments["config"] as? String
        val configLen = config?.length ?: 0
        return "method=connect keys=[$keys] config_len=$configLen"
    }
}
